### PR TITLE
TELCODOCS-1785 Improving NROP documentation

### PR DESCRIPTION
**Pull request by kquinn1204**
[TELCODOCS-1785]: Improving NROP documentation

Version(s): 4.12 > 

Issue: https://issues.redhat.com/browse/TELCODOCS-1785

Link to docs preview: https://73634--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-numa-aware-scheduling

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

**Comment by kquinn1204** [comment]
@shajmakh please review this update to the documentation associated with NROP

File: modules/cnf-scheduling-numa-aware-workloads.adoc [hunk 1]
```diff
@@ -30,7 +30,7 @@ $ oc get numaresourcesschedulers.nodetopology.openshift.io numaresourcesschedule
 .Example output
 [source,terminal]
 ----
```
Comment (shajmakh) [inline modules/cnf-scheduling-numa-aware-workloads.adoc:32]: a bit earlier in this section, perhaps in the entry, let's clearly add that "...Now that topology-aware scheduling operator is installed, numaresourcesoperator and numaresourcesschedulers CRs are applied and your cluster has a matching performanceprofile or kubeletconfig, (next sentence is taken from the current documentation for this section) You can schedule workloads with the NUMA-aware scheduler using Deployment CRs that specify the minimum required resources to process the workload."

**Review by shajmakh — COMMENTED** [review]
Thanks @kquinn1204 for your efforts handling this. 
added some comments below. 

There is also an update needed for all versions which is that podsFingerprinting parameter now can accept 3 values: Enabled, Disabled, and EnabledExclusiveResources (missing in docs).

The next update is that podsFingerprinting default value in releases <4.16 is Enabled while in 4.16 it has changed to EnabledExclusiveResources. let's convey this detail in the docs too.

File: modules/cnf-creating-nrop-cr.adoc [hunk 2]
```diff
@@ -53,5 +55,5 @@ $ oc get numaresourcesoperators.nodetopology.openshift.io
 [source,terminal]
 ----
 NAME                    AGE
-numaresourcesoperator   10m
+numaresourcesoperator   27s
 ----
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-creating-nrop-cr.adoc:59]: I will have to create another PR for this "The next update is that `podsFingerprinting` default value in releases <4.16 is Enabled while in 4.16 it has changed to `EnabledExclusiveResources`. let's convey this detail in the docs too." as 4.16 will not be released before I get published. I will create a new PR for that update.

File: modules/cnf-configuring-single-numa-policy.adoc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    kubeReserved:
+      memory: "512Mi"
+    reservedMemory:
+      - numaNode: 0
+        limits:
+          memory: "1124Mi"
+    systemReserved:
+      memory: "512Mi"
+    topologyManagerPolicy: "single-numa-node" <5>
+    topologyManagerScope: "pod"
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-configuring-single-numa-policy.adoc:143]: Hi @shajmakh do I just remove , I ma bit confused what does the kubeconfig need to look like to ensure NRO is set up correct;y ?

```
systemReserved:
      memory: "512Mi"
    topologyManagerPolicy: "single-numa-node" <5>
    topologyManagerScope: "pod"
```

File: modules/cnf-configuring-single-numa-policy.adoc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+<4> For `memoryManagerPolicy`, `Static` must use an uppercase `S`.
+<5> `topologyManagerPolicy` must be set to `single-numa-node`.
+
+.. Create the `KubeletConfig` custom resource (CR) by running the following command:
++
+[source,terminal]
+----
+$ oc create -f nro-kubeletconfig.yaml
+----
+
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-configuring-single-numa-policy.adoc:157]: Is manual intervention required to reboot the nodes or is that automatic ?

[the remark this answers is not in the document]
Comment (shajmakh) [reply modules/cnf-configuring-single-numa-policy.adoc]: it is automatic
if no reboot is triggered something is missing and the user should troubleshoot that by looking into the labels in KC that address the node group

File: modules/cnf-configuring-single-numa-policy.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (shajmakh) [reply modules/cnf-configuring-single-numa-policy.adoc]: sorry for confusion, I meant to mark one line only, let's just remove the line that specifies topologyManagerScope

**Comment by shajmakh** [comment]
@kquinn1204 thank you for your efforts on this. Please ping me when the PR is ready for another review round.

**Comment by kquinn1204** [comment]
@shajmakh hopefully good to go now. Thanks for your input.

File: modules/cnf-deploying-the-numa-aware-scheduler.adoc [hunk 5]
```diff
@@ -88,16 +31,7 @@ spec:
...
-
-. Verify that the performance profile was applied by running the following command:
-+
-[source,terminal]
-----
-$ oc describe performanceprofile <performance-profile-name>
-----
-
-. Verify that the required resources deployed successfully by running the following command:
+. After a few minutes run the following command to verify that the required resources deployed successfully:
```
Comment (shajmakh) [inline modules/cnf-deploying-the-numa-aware-scheduler.adoc:34]: applying the scheduler doesn't trigger reboot so no need to wait few mins, it is just matter of secs

File: modules/cnf-sample-single-numa-policy-from-pp.adoc [hunk 6]
```diff
@@ -0,0 +1,37 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_module-type: REFERENCE
+[id="cnf-sample-single-numa-policy_{context}"]
+= Sample single NUMA node policy
+
+An example performance profile created using the PPC tool is shown here:
```
Comment (shajmakh) [inline modules/cnf-sample-single-numa-policy-from-pp.adoc:9]: An example _of_ performance profile ..

File: modules/cnf-sample-single-numa-policy-from-pp.adoc [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+----
+---
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: performance
+spec:
+  cpu:
+    isolated: "3"
+    reserved: 0-2
```
Comment (shajmakh) [inline modules/cnf-sample-single-numa-policy-from-pp.adoc:21]: let's also note to the user to adjust the cpus as needed by their machines

File: modules/cnf-sample-single-numa-policy-from-pp.adoc [hunk 8]
```diff
@@ -0,0 +1,37 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_module-type: REFERENCE
+[id="cnf-sample-single-numa-policy_{context}"]
+= Sample single NUMA node policy
```
Comment (shajmakh) [inline modules/cnf-sample-single-numa-policy-from-pp.adoc:7]: the sample is for performane profile

**Comment by shajmakh** [comment]
@kquinn1204 thanks for addressing the comments. 
some nits and we should be good to go.

File: modules/cnf-sample-single-numa-policy-from-pp.adoc (hunk 6 above)
Comment (kquinn1204) [reply modules/cnf-sample-single-numa-policy-from-pp.adoc:9]: rephrased it slightly to be in active voice

File: modules/cnf-scheduling-numa-aware-workloads.adoc [hunk 9]
```diff
@@ -131,25 +130,25 @@ Deployments that request more resources than is available for scheduling will fa
 
 . Verify that the expected allocated resources are listed for the node.
 
-.. Identify the node that is running the deployment pod by running the following command, replacing <namespace> with the namespace you specified in the `Deployment` CR:
+.. Identify the node that is running the deployment pod by running the following command, the namespace is that specified in the `Deployment` CR:
```
Comment (kcarmichael08) [inline modules/cnf-scheduling-numa-aware-workloads.adoc:133]:
```suggestion
.. Identify the node that is running the deployment pod by running the following command, including the namespace is that specified in the `Deployment` CR:
```
I think there is a word missing here

File: modules/cnf-scheduling-numa-aware-workloads.adoc [hunk 10]
```diff
@@ -131,25 +130,25 @@ Deployments that request more resources than is available for scheduling will fa
...
 .Example output
 [source,terminal]
 ----
 NAME                                 READY   STATUS    RESTARTS   AGE   IP            NODE     NOMINATED NODE   READINESS GATES
-numa-deployment-1-65684f8fcc-bw4bw   0/2     Running   0          82m   10.128.2.50   worker-0   <none>  <none>
+numa-deployment-1-6c4f5bdb84-wgn6g   0/2     Running   0          82m   10.128.2.50   worker-1   <none>  <none>
 ----
 +
-.. Run the following command, replacing <node_name> with the name of that node that is running the deployment pod.
+.. Run the following command, with the name of that node that is running the deployment pod.
```
Comment (kcarmichael08) [inline modules/cnf-scheduling-numa-aware-workloads.adoc:147]:
```suggestion
.. Run the following command with the name of that node that is running the deployment pod.
```

**Review by kcarmichael08 — COMMENTED** [review]
Wow, that was a lot of info - nicely organized! Had some comments and saw a few typos. Nice job!

File: modules/cnf-about-numa-aware-scheduling.adoc [hunk 11]
```diff
@@ -6,15 +6,17 @@
 [id="cnf-about-numa-aware-scheduling_{context}"]
 = About NUMA-aware scheduling
 
-Non-Uniform Memory Access (NUMA) is a compute platform architecture that allows different CPUs to access different regions of memory at different speeds. NUMA resource topology refers to the locations of CPUs, memory, and PCI devices relative to each other in the compute node. Co-located resources are said to be in the same _NUMA zone_. For high-performance applications, the cluster needs to process pod workloads in a single NUMA zone.
+Introduction to NUMA:: Non-Uniform Memory Access (NUMA) is a compute platform architecture that allows different CPUs to access different regions of memory at different speeds. NUMA resource topology refers to the locations of CPUs, memory, and PCI devices relative to each other in the compute node. Co-located resources are said to be in the same _NUMA zone_. For high-performance applications, the cluster needs to process pod workloads in a single NUMA zone.
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-about-numa-aware-scheduling.adoc:9]: I was trying to address the wall of text and break it up. Let me look at discrete headers thanks for the heads up excuse the pun

File: modules/cnf-configuring-kubelet-nro.adoc [hunk 12]
```diff
@@ -0,0 +1,60 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_module-type: PROCEDURE
+[id="cnf-configuring-kubelet-config-nro_{context}"]
+= Creating a KubeletConfig CRD
+
+An alternative way to configure a single NUMA node policy is to create and apply a `KubeletConfig` custom resource (CR) by using the following procedure:
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-configuring-kubelet-nro.adoc:9]: I guess it was following on from "There are two ways to achieve this by creating and applying a performance profile or a KubeletConfig." but may be this earlier reference is hard to keep track of here.

File: modules/cnf-nrop-with-mcp.adoc
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-nrop-with-mcp.adoc:1]: sorry was a draft think I need to delete

File: modules/cnf-nrop-with-mcp.adoc [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+spec:
+  machineConfigSelector:
+    matchLabels:
+      machineconfiguration.openshift.io/role: worker
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+----
+
+.. Create the `MachineConfigPool` CR by running the following command:
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/cnf-nrop-with-mcp.adoc:38]: think I should remove

File: modules/cnf-configuring-kubelet-nro.adoc [hunk 14]
```diff
@@ -0,0 +1,60 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_module-type: PROCEDURE
+[id="cnf-configuring-kubelet-config-nro_{context}"]
+= Creating a KubeletConfig CRD
+
+The recommended way to configure a single NUMA node policy is to apply a performance profile. Another way is by creating and applying a `KubeletConfig` custom resource (CR), as shown in the following procedure:
```
Comment (stevsmit) [inline modules/cnf-configuring-kubelet-nro.adoc:9]:
```suggestion
The recommended way to configure a single NUMA node policy is to apply a performance profile. Another way is by creating and applying a `KubeletConfig` custom resource (CR), as shown in the following procedure.
```

File: modules/cnf-about-numa-aware-scheduling.adoc [hunk 15]
```diff
@@ -6,13 +6,44 @@
...
+[discrete]
+[id="performance-considerations_{context}"]
+== Performance considerations
+
+NUMA architecture allows a CPU with multiple memory controllers to use any available memory across CPU complexes, regardless of where the memory is located. This allows for increased flexibility at the expense of performance. A CPU processing a workload using memory that is outside its NUMA zone is slower than a workload processed in a single NUMA zone. Also, for I/O-constrained workloads, the network interface on a distant NUMA zone slows down how quickly information can reach the application. High-performance workloads, such as telecommunications workloads, cannot operate to specification under these conditions. 
+
+[discrete]
+[id="numa-aware-scheduling_{context}"]
+
+== NUMA-aware scheduling
```
Comment (stevsmit) [inline modules/cnf-about-numa-aware-scheduling.adoc:24]: Backspace so that it's under the ID

File: modules/cnf-about-numa-aware-scheduling.adoc [hunk 16]
```diff
@@ -6,13 +6,44 @@
...
+
+NUMA-aware scheduling aligns the requested cluster compute resources (CPUs, memory, devices) in the same NUMA zone to process latency-sensitive or high-performance workloads efficiently. NUMA-aware scheduling also improves pod density per compute node for greater resource efficiency.
+
+[discrete]
+[id="integration-with-node-tuning-operator_{context}"]
+
+== Integration with Node Tuning Operator
```
Comment (stevsmit) [inline modules/cnf-about-numa-aware-scheduling.adoc:31]: Backspace so that it's under the ID

File: modules/cnf-about-numa-aware-scheduling.adoc [hunk 17]
```diff
@@ -6,13 +6,44 @@
...
 
-The default {product-title} pod scheduler scheduling logic considers the available resources of the entire compute node, not individual NUMA zones. If the most restrictive resource alignment is requested in the kubelet topology manager, error conditions can occur when admitting the pod to a node. Conversely, if the most restrictive resource alignment is not requested, the pod can be admitted to the node without proper resource alignment, leading to worse or unpredictable performance. For example, runaway pod creation with `Topology Affinity Error` statuses can occur when the pod scheduler makes suboptimal scheduling decisions for guaranteed pod workloads by not knowing if the pod's requested resources are available. Scheduling mismatch decisions can cause indefinite pod startup delays. Also, depending on the cluster state and resource allocation, poor pod scheduling decisions can cause extra load on the cluster because of failed startup attempts.
+[discrete]
+[id="default-scheduling-logic_{context}"]
+
+== Default scheduling logic
```
Comment (stevsmit) [inline modules/cnf-about-numa-aware-scheduling.adoc:38]: Backspace so that it's under the ID

File: modules/cnf-configuring-kubelet-nro.adoc [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    reservedMemory:
+      - numaNode: 0
+        limits:
+          memory: "1124Mi"
+    systemReserved:
+      memory: "512Mi"
+    topologyManagerPolicy: "single-numa-node" <5>
+----
+<1> Adjust this label to match the the `machineConfigPoolSelector` in the `NUMAResourcesOperator` CR.
+<2> Adjust this based on the CPU on your nodes.
```
Comment (stevsmit) [inline modules/cnf-configuring-kubelet-nro.adoc:45]: There isn't something right with the numbering here. 

    cpuManagerPolicy: "static" <2>

however, it's description is:
<2> Adjust this based on the CPU on your nodes.

Based on the callouts, it look like     cpuManagerPolicy: "static" <2> should be <3> For `cpuManagerPolicy`, `static` must use a lowercase `s`.

File: modules/cnf-configuring-single-numa-policy.adoc [hunk 19]
```diff
@@ -0,0 +1,16 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_module-type: PROCEDURE
+[id="cnf-configuring-single-numa-policy_{context}"]
+= Configuring a single NUMA node policy
+
+The NUMA Resources Operator needs a single NUMA node policy configured on the cluster. There are two ways to achieve this by creating and applying a performance profile or a `KubeletConfig`.
```
[the remark this answers is not in the document]
Comment (stevsmit) [reply modules/cnf-configuring-single-numa-policy.adoc:9]: @kquinn1204 suggest adding this colon

File: modules/cnf-nrop-with-mcp.adoc [hunk 20]
```diff
@@ -0,0 +1,43 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
+
+:_module-type: PROCEDURE
+[id="cnf-nrop-creating-mcp-cr_{context}"]
+= Creating a MachineConfigPool custom resource
+
+By default, {product-title} creates a MachineConfigPool resource for worker nodes in the cluster. To tailor configurations for specific workloads, you should create a custom MachineConfigPool resource. Follow this procedure to define a pool named `worker-cnf` for telecommunications workloads..
```
Comment (stevsmit) [inline modules/cnf-nrop-with-mcp.adoc:9]:
```suggestion
By default, {product-title} creates a MachineConfigPool resource for worker nodes in the cluster. To tailor configurations for specific workloads, you should create a custom MachineConfigPool resource. Follow this procedure to define a pool named `worker-cnf` for telecommunications workloads. 
```

File: modules/cnf-nrop-with-mcp.adoc [hunk 21]
```diff
@@ -0,0 +1,43 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-numa-aware-scheduling.adoc
```
Comment (stevsmit) [inline modules/cnf-nrop-with-mcp.adoc:3]: I do not see this module in the preview.

File: modules/cnf-sample-single-numa-policy-from-pp.adoc [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+:_module-type: REFERENCE
+[id="cnf-sample-performance-policy_{context}"]
+= Sample performance profile
+
+This example YAML shows a performance profile created by using the performance profile creator (PPC) tool: 
+
+[source,yaml]
+----
+---
```
Comment (stevsmit) [inline modules/cnf-sample-single-numa-policy-from-pp.adoc:13]:
```suggestion

```

**Comment by stevsmit** [comment]
@kquinn1204 A few more nits. Most wouldn't have held up a merge review, but the YAML issue I pointed out needs to be fixed before merge. Also, the rogue module needs deleted if it isn't going to be used. ping when changes are made and I can merge it.

File: modules/cnf-configuring-kubelet-nro.adoc (hunk 18 above)
Comment (kquinn1204) [reply modules/cnf-configuring-kubelet-nro.adoc:45]: good catch thanks

File: modules/cnf-nrop-with-mcp.adoc (hunk 20 above)
Comment (kquinn1204) [reply modules/cnf-nrop-with-mcp.adoc:9]: removed extra now not needed

**Comment by kquinn1204** [comment]
Thank you @stevsmit hopefully should be good to merge now.

**Pull request merged**